### PR TITLE
Hierarchical area selection

### LINKED ISSUE #3: Use cascading selectors for area selection
Travis suggests cascading selectors for areas selection, perhaps `Basin` -> `Watershed` -> `Conservation Unit.` 

Users would select a `Basin`, then see the `Basin` highlighted on the map and displayed in the data. They could then look at `Basin` data, or move to the next dropdown menu, which would show a list of `Watershed`s. If they selected a `Watershed`, the `Watershed` would be highlighted on the map and displayed in the data. They could look at that, or go to the final dropdown, which would be a list of the `Conservation Unit`s in that watershed.If they selected a `Conservation Unit`, they would see it highlighted on the map and displayed in the data.

Travis has provided information on which watersheds are in which Basin, and which Conservation Units are in which Watersheds, but in order to implement this, I would also need basin and conservation unit boundaries (shapefiles to serve with geoserver).

Definitely seems like it would be an improvement over the current "selector with hundreds of entries" UI.

### PR DESCRIPTION
**Pull request by corviday**
Allows users to navigate area selection via hierarchical dropdowns. 

There is now a "basin" dropdown, with which a user can select a river basin like "Fraser" or "Nass". This has two effects: it will highlight the entire basin on the map and display salmon population data for the entire basin, and it will populate the conservation unit and watershed dropdowns with conservation units or watersheds that are inside that basin.

Demo [here](https://services.pacificclimate.org/dev/scip/app/).

**Possibly confusing data limitations that will show up in the demo:**

At present, we only have indicator/variable data for the Fraser Basin. Salmon population data is available for all watersheds, but graphs and stream networks are only available within the Fraser. Additionally, watersheds are not defined according to stream networks, so the stream network of a selected watershed may extend outside the watershed.

Also some regions have odd names containing brackets; these were present in the shapefiles I received and I do not know whether to update them.

Conservation unit and watershed shapes are from different data sources, and sometimes "adjacent" regions are treated as nesting; fixing this probably involved updating the backend to accepts a minimum overlapping area term.

Future work:
This is part one of a two- or three-part overhaul of area selection. Part two will allow users to further narrow down the list of possible watersheds or conservation units by specifying one or more salmon species they are interested in. 

resolves #3

### CODE REVIEW

**Comment by rod-glover** [comment]
Question: Are the watersheds supposed to be entirely inside the basins? Or is just a bit of overlap that makes it elegible?

Example:
1. Select Fraser River basin.
2. Select Babine Lake watershed. It has only a very small overlap with Fraser basin. At far zoom outs, it looks like no overlap.

**Comment by corviday** [comment]
Watersheds _should_ be entirely inside basins, but obviously aren't. I believe the correct fix for this involves changing the backend defines with "overlaps" - that is, setting some minimum area required in order for two shapes to be considered overlapping.

**Comment by rod-glover** [comment]
Ah, that makes more sense. Maybe a minimum fraction of area of overlap. What the denominator should be is an interesting question: minimum of the two? sum of the two?

**Review by rod-glover — CHANGES_REQUESTED** [review]
This works very nicely! It looks as if the Basin and Species selectors do what they're supposed to.

Some broader questions/comments (maybe to be addressed elsewhere):
1. When are the other river basins going to receive data? It might be good to disable them in the Basin dropdown until they do have data. It's quite confusing for a naive user (that's me) to select a basin with no data - the white outline is very hard to spot against only the base map.
2. Did you say you were having trouble with layout of the checkboxes? It looks like you may. If so, may I suggest letting React Bootstrap do the layout work for you? Specifically, you can use a[ `<ButtonToolbar>`](https://react-bootstrap.github.io/docs/components/button-group) to enclose a set of [`<Form.Check type="checkbox" />` ](https://react-bootstrap.github.io/docs/forms/checks-radios)items. Example:
```
    <ButtonToolbar aria-label="Basic example">
      {[1,2,3,4,5,6].map(i => (<Form.Check type={"checkbox"} id={`foo`} label={`Your amazing salmon species ${i}`}  className="me-2"/>))}
    </ButtonToolbar>
```
This lays out the checkboxes horizontally (for some reason `ButtonGroup` doesn't, but this does.) The `me-2` is a [helper class](https://getbootstrap.com/docs/4.0/utilities/spacing/) that adds margins to items it's applied to, so the checkboxes are spaced nicely. You could choose a different helper class. 

Of course you'll map the `taxons` array, not integers.

I think it would be worth discussing these things:
1. When and where to introduce state in a React component.
2. When to use and not use `useEffect`.
3. Computing values derived from props and state inside a component. It's actually pleasingly simple.

File: src/components/AreaDisplay/AreaDisplay.js [hunk 1]
```diff
@@ -2,42 +2,221 @@
...
+      else {
+        newSelectedTaxons = [];
+        for(let i = 0; i < selectedTaxons.length; i++) {
+            if(selectedTaxons[i].common_name !== taxon.common_name || selectedTaxons[i].subgroup !== taxon.subgroup) {
+                newSelectedTaxons.push(selectedTaxons[i]);
+            }
+        }
+      }
+      setSelectedTaxons(newSelectedTaxons);
+  }
```
[the remark this answers is not in the document]
Comment (rod-glover) [reply src/components/AreaDisplay/AreaDisplay.js:101]: This is an optional change. More FYI than a change request.

**Comment by corviday** [comment]
>     1. When are the other river basins going to receive data? It might be good to disable them in the Basin dropdown until they do have data. It's quite confusing for a naive user (that's me) to select a basin with no data - the white outline is very hard to spot against only the base map.

Data for coastal regions is very nearly ready. Data for the interior might be a year off.

File: src/components/TaxonSelector/TaxonCheckbox.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    useEffect(() => {
+        setChecked(_.find(selectedTaxons, taxon, _.isEqual) ? true : false); 
+      }, [taxon, selectedTaxons]
+    );    
+    
+    const handleChange = () => {
+        onChange(taxon, !checked);
+        setChecked(!checked);
+    };
+
```
[the remark this answers is not in the document]
Comment (corviday) [reply src/components/TaxonSelector/TaxonCheckbox.js:22]: ah, yes, thank you. That's much better!

**Comment by corviday** [comment]
> Did you say you were having trouble with layout of the checkboxes? It looks like you may. If so, may I suggest letting React Bootstrap do the layout work for you? Specifically, you can use a[ <ButtonToolbar>](https://react-bootstrap.github.io/docs/components/button-group) to enclose a set of [<Form.Check type="checkbox" /> ](https://react-bootstrap.github.io/docs/forms/checks-radios)items.

I am having layout troubles! Can I get your opinion on these options, various combinations of your suggestions of `ButtonToolbar` and `me-2`:

current layout, as a reminder:
current

Changing to `ButtonToolbar` and `me-2` and keeping everything else constant - doesn't quite work because the labels are too long:
toolbar

Using `ButtonToolbar` and `me-2`, and putting the taxon checkboxes in their own Row, instead of sharing with the basin dropdown:
taxonrow

Using only `me-2` to keep the current layout, but with nice spacing:
me-2only

I think I prefer the last option, but would like your opinion. Is there something else I should consider for this layout?

**Comment by rod-glover** [comment]
I like the last option best. It also looks like you've put each item into a <Col>, which gives a nice lineup with the row below.

Side note: If you want (not suggesting you need it), you can increase the spacing between checkboxes by using `me-3` or higher. There is an abundance of nice helper classes in Bootstrap, and I love it when I realize I can just use them instead of crafting my own. Plus they always work right!

**Comment by corviday** [comment]
Yeah! I tried adding margins with CSS, but it didn't work; I'm sure I was doing something wrong. me-2 is _easy_!

File: src/components/AreaDisplay/AreaDisplay.js [hunk 3]
```diff
@@ -2,42 +2,221 @@
...
+            data => {
+                setWatersheds(parseRegions(data));
+            }
+        );
+        getConservationUnits(boundary).then(
+            data => {
+                setConservationUnits(parseRegions(data));
+            }
+        );
+      }
```
[the remark this answers is not in the document]
Comment (corviday) [reply src/components/AreaDisplay/AreaDisplay.js:131]: Yes, but the backend "give me all watersheds" call is _significantly_ faster and easier than making seven "give me chum watersheds, give me coho watersheds, etc" calls and merging the results - not only is the latter seven separate calls with a fair bit of overlap in the results, but each call contains a spatial join between the region table where watershed geometries are and the conservation unit table of where salmon are.

The empty list case is redundant, though, I'll get rid of that one and document the reasoning on the other.

**Comment by corviday** [comment]
I have improved the formatting on the checkboxes and fixed the React lifecycle issues.

File: src/components/AreaDisplay/AreaDisplay.js [hunk 4]
```diff
@@ -2,42 +2,217 @@
...
+  // contain the currently selected species are available.
 
-  //fetch region list from the API.
+  // fetch basin list from the API.
   // this only needs to be done once, when the component is loaded
   useEffect(() => {
-    if(regions.length === 0) {
-        getWatersheds().then(
+      if(basins.length === 0) {
+          getBasins().then(
```
Comment (rod-glover) [inline src/components/AreaDisplay/AreaDisplay.js:55]: Possible bug here: If the backend returns a zero-length list, then you enter an infinite loop of requests. Unlikely but not strictly speaking impossible.

File: src/components/TaxonSelector/TaxonCheckbox.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+function TaxonCheckbox({taxon, selectedTaxons, onChange}) {
+    
+    const checked = _.find(selectedTaxons, taxon, _.isEqual) ? true : false;
+    
+    const handleChange = () => {
+        onChange(taxon, !checked);
+    };
+    
+    return (
+//        <label>
```
Comment (rod-glover) [inline src/components/TaxonSelector/TaxonCheckbox.js:20]: Remove commented out lines?

File: src/helpers/GeographyHelpers.js [hunk 6]
```diff
@@ -1,23 +1,20 @@
...
-export function filterRegions(regions) {
-    return _.filter(regions, region => {
-    return(_.includes(upper_fraser_watersheds, region.name)); 
-    })
+// Is point a valid WKT point?
+export function validPoint(point) {
+    const p = JSON.parse(point)
+    return p 
+        && _.has(p, 'type')
+        && p.type === "Point"
```
Comment (rod-glover) [inline src/helpers/GeographyHelpers.js:13]: You can squish the previous three lines down to `p?.type === "Point"`, using the [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) operator.

File: src/helpers/GeographyHelpers.js [hunk 7]
```diff
@@ -1,23 +1,20 @@
...
+        && _.has(p, 'coordinates') 
+        && p.coordinates.length === 2 
+        && _.isNumber(p.coordinates[0])
+        && _.isNumber(p.coordinates[1]);
```
Comment (rod-glover) [inline src/helpers/GeographyHelpers.js:17]: And the series of `&&`s here can become `_.isNumber(p?.coordinates?.[0]) && _.isNumber(p?.coordinates?.[1])`

**Review by rod-glover — APPROVED** [review]
LGTM! A few minor items for you to consider, but all good.

Some observations/questions on the UI:
- When you change the species selection, the previous watershed or conservation unit selection remains, even if it is invalid. (Easiest way to see this is to select a watershed, deselect all species; watershed is still selected even tho the option list is now empty.) Probably changing species should also remove those selections.
- I think a Select All and Clear All control for the species would be helpful. You could use buttons, or that trick where a single checkbox is used, but it might be trickier to implement.
- I wonder if it might be better to place the dropdown labels on the same line as the dropdowns, at least for larger layouts. Or maybe you could use the lovely [Floating labels](https://react-bootstrap.github.io/docs/forms/floating-labels) feature of React Bootstrap. I mean particularly for the basin, watershed, and conservation unit selectors, although it applies equally to all drop
- (I know this isn't directly related to this PR, but this should probably go into an issue to be addressed.) On the Yearly Indicators tab, there is no default value for Climate Model or Emissions Scenario, yet when you select a Variable, you see a graph. What model and emissions are used for that graph? It should show up in the selectors. Same for Monthly Indicators, Daily Indicators, Salmon Populations.

**Comment by corviday** [comment]
I was hoping to implement some of the excellent UI suggestions "quickly" but it took long enough I think it belongs in a separate PR, so I have just done the chaining operator suggestions. I didn't know about that, thanks!

**Pull request merged**